### ISSUE TITLE
[Bug][4.3.0] Oauth authentication against gitlab does not work any more

### ISSUE THREAD

**Issue by do3cc** [grafana/grafana#8460]
We are using Grafana 4.3.1. We tested against Grafana 4.3.0 and Grafana 4.2.0 too
Our Auth datasource is gitlab. We verified the configuration by trying it with Grafana 4.2.0, where authentication is possible.

oauth configuration:
```
[auth.generic_oauth]
enabled = true
name = GitLab
allow_sign_up = true
client_id = 1234567...
client_secret = 1234567...
scopes = read_user openid
auth_url = https://domain/oauth/authorize
token_url = https://domain/oauth/token
api_url = https://domain/api/v4/user
;team_ids =
;allowed_organizations =
```

I tried to log in with oauth
I expected to be logged in.
Instead I get an error in the browser:
 Server side error :(
Failed to create account

In the logs with loglevel trace I get:
```
May 24 16:23:45 sp1 grafana-server[1095]: t=2017-05-24T16:23:45+0200 lvl=dbug msg="Scheduling update" logger=alerting.scheduler ruleCount=1
May 24 16:23:48 sp1 grafana-server[1095]: t=2017-05-24T16:23:48+0200 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/login/generic_oauth status=302 remote_addr=11.12.13.14 time_ms=0 size=344
May 24 16:23:48 sp1 grafana-server[1095]: t=2017-05-24T16:23:48+0200 lvl=dbug msg="OAuthLogin Got token" logger=context userId=0 orgId=0 uname=
May 24 16:23:48 sp1 grafana-server[1095]: t=2017-05-24T16:23:48+0200 lvl=dbug msg="OAuthLogin got user info" logger=context userId=0 orgId=0 uname= userInfo="&{Name: Email: Login: Company: Role:}"
May 24 16:23:48 sp1 grafana-server[1095]: t=2017-05-24T16:23:48+0200 lvl=eror msg="Failed to create account" logger=context userId=0 orgId=0 uname= error="pq: doppelter Schlüsselwert verletzt Unique-Constraint »UQE_user_login«"
May 24 16:23:48 sp1 grafana-server[1095]: t=2017-05-24T16:23:48+0200 lvl=eror msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/login/generic_oauth status=500 remote_addr=11.12.13.14 time_ms=58 size=688
```
I did NOT sanitize the userIds and orgIds, only the IPs.
Sorry for the german error message in there but I think it should be clear that an unique constraint is failing there.
I found a fishy user in sql:
```
-bash-4.2$ psql -d grafana -c 'select * from public.user order by id DESC limit 1'
 id | version | login | email | name | password | salt | rands | company | org_id | is_admin | email_verified | theme |       created       |       updated       | help_flags1
----+---------+-------+-------+------+----------+------+-------+---------+--------+----------+----------------+-------+---------------------+---------------------+-------------
 25 |       0 |       |       |      |          |      |       |         |      1 | f        | f              |       | 2017-05-24 14:37:11 | 2017-05-24 14:37:11 |           0
(1 row)
```
Deleting this row, I can log in once and a bad entry like above is in the table again
I looked at the changes in the relevant files from 4.2.0 to 4.3.0 but did not find anything obvious and I am out of ideas for now.

**Comment by torkelo** [comment]
what oauth options are you using?

**Comment by do3cc** [comment]
I've updated the ticket with that part of the configuration. As written above, we verified with 4.2.0 that the configuration is working.

**Comment by DanCech** [comment]
This may be related to the changes to support bitbucket style oauth endpoints, I'll see if i can duplicate the issue against gitlab.com

**Comment by DanCech** [comment]
If you change your config to `scopes = read_user openid api` it will work.  Without `api` gitlab returns `{"error":"insufficient_scope","error_description":"The request requires higher privileges than provided by the access token.","scope":"api"}`

I will look into why the request isn't being marked as failed in this situation, even though Gitlab is returning a 403 http response code.

**Comment by do3cc** [comment]
@DanCech you are right, this did the trick. I also tried this two days ago but must have mistakenly not restarted grafana while testing.
I did not see the error you mentioned.
Should I close this ticket or do you want to use it to keep track of the error reporting?

**Comment by DanCech** [comment]
@do3cc Good to hear that solved the problem 👍 The error handling update was merged in #8470 so we should be all set.
